### ISSUE TITLE
A working example of FreeIPA in Kubernetes 1.26

### ISSUE THREAD

**Issue by iPenguin** [freeipa/freeipa-container#529]
Hi,

After searching for current examples of working Kubernetes manifests that I could use to deploy FreeIPA in my home lab I ended up having to piece one together from a number of posts and repos. I thought it would be good for me to document what I learned.

I'm [deploying FreeIPA as a StatefulSet](https://github.com/iPenguin/bare-metal-kubernetes/blob/master/roles/freeipa/templates/freeipa-statefulset.yaml.j2) with a MetalLB [LoadBalancer in front of it](https://metallb.universe.tf/](https://github.com/iPenguin/bare-metal-kubernetes/blob/master/roles/freeipa/templates/freeipa-lb.yaml.j2)).

The biggest thing I found missing in all the examples I saw was the need for a `startupProbe`. This block prevents the pod from being killed during the long startup on initial install (15+ minutes) without adversely impacting pod restarts.

```
  startupProbe:
    httpGet:
      path: /
      port: 80
    failureThreshold: 35
    periodSeconds: 30
``` 

I hope someone else finds this helpful.
-Brian

**Comment by adelton** [comment]
Killing of the pod with just the `readinessProbe` was mentioned in https://github.com/freeipa/freeipa-container/issues/516#issuecomment-1434803180 but I never actually saw it happen. The document at https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/ says
> For example, an application might need to load large data or configuration files during startup, or depend on external services after startup. In such cases, you don't want to kill the application, but you don't want to send it requests either. Kubernetes provides readiness probes to detect and mitigate these situations.

Can you check https://github.com/freeipa/freeipa-container/blob/master/tests/freeipa-k3s.yaml (potentially with `freeipa-k3s-yaml-*.patch`es from https://github.com/freeipa/freeipa-container/tree/master/tests applied) to see if you get the pod/container killed with just this Pod definition?

**Comment by adelton** [comment]
I also wonder why you went with a StatefulSet with `replicas: 1`, rather than just a Pod? Have you figured a way to automatically autocreate FreeIPA replicas with ` replicas` value greater than 1?

**Comment by iPenguin** [comment]
Hi @adelton,

**TL;DR**
I tested using only the example `readinessProbe` in a pod and it looks like that works, but any reasonable `livenessProbe` will kill the install process so either remove it or add a `startupProbe`. `StatefulSets` have life-cycle management benefits over standalone pods.

**Longer Version**
When I started building my manifest and it started the installation process the `livenessProbe` would eventually kill the pod because the install was taking so long and the services were not yet ready. 

I experimented with setting the probes to really long `initialDelaySeconds` (1200 seconds) and everything would install and I'd get a running instance. However, if I deleted the pod, the initial delay in the probes would mean the new pod would not register as ready for 1200 seconds and that delay meant the LoadBalancer wouldn't point to the new pod for 20 minutes and IPA would be inaccessible for that time.

In order to get a better understanding of what the probes were doing I did some research and came across the link you shared. There is a section in the article on [Protecting slow starting containers with startup probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) which is how I came to my solution to add a `startupProbe` and leave the liveness and readiness probes with more reasonable numbers.

I'm using a `StatefulSet` because they are designed for applications like FreeIPA where each instance is unique.  They also have some nice life-cycle management features that a standalone pod doesn't have. For example, with a `StatefulSet` you can delete the pod and Kubernetes will see that the `StatefulSet` doesn't have the right number of pods and create a new one automatically.

You are correct that a `StatefulSet` will not give me the ability to scale up replicas of FreeIPA. I haven't looked into that yet, but I suspect the solution may involve having a different `StatefulSet` for each instance of IPA.

In reference to trying the IPA example manifests: I tried to run it, but it didn't work, and I didn't want to spend the time trying to figure out all the issues. So instead I started with my manifest, I removed the `StatefulSet`, leaving just the pod and replaced the probes with just the `readinessProbe` from the example. With those changes I was able to get a working instance of IPA. If you don't care about a liveness probe you could remove that instead of adding a `startupProbe`.

-Brian

**Comment by abbra** [comment]
@iPenguin did you look at https://github.com/freeipa/freeipa-openshift-container and https://github.com/freeipa/freeipa-operator ? While we don't work on those projects anymore, they have been looking at solving the same issues.

**Comment by adelton** [comment]
@abbra If there are some specific take-aways or improvements from the work on https://github.com/freeipa/freeipa-openshift-container and https://github.com/freeipa/freeipa-operator which could be added to https://github.com/freeipa/freeipa-container's README or for example the https://github.com/freeipa/freeipa-container/blob/master/tests/run-master-in-k3s.sh, it might be good to get them in before the knowledge fades. We actually had two issues opened on the same day talking about the StatefulSet approach (this issue and https://github.com/freeipa/freeipa-container/issues/528) so it seems folks have made some progress in their K8s deployments.

**Comment by iPenguin** [comment]
Hi @abbra,

Yes I had come across those repos when I was doing research. Other then looking through them for clues, I didn't do too much with them as the were both abandoned, and the operator was listed as experimental.

The fact that I was able to get a running instance of IPA is really due to the fact that I was able to pull in details from many different examples that are available, so thank you everyone! I just wanted to contribute the little bit I pieced together on my own.

Thanks,
-Brian

**Comment by 1nfr4M4vrk** [comment]
Hello,
First of all, thank you so much for your great and active community!

I'm writing in reference to the manifest that user @iPenguin  kindly shared earlier.
 I tried deploying the manifest myself, and I encountered a problem during restart named service in install phase — it fails unless I set privileged: true.

When I do set privileged: true, everything works well and the installation completes successfully. However, after a pod restart, the hostname changes, and the FreeIPA service fails to start correctly.

I checked the logs of the named service, and it fails with a systemd start service timeout, which seems to be the root cause

Here are a few additional details about my environment:

    Kubernetes version: 1.29.5

    OS: Rocky Linux 9.4

    SELinux: Disabled

I’ve tried the following to troubleshoot:

    Used different base images like Rocky and CentOS.

    Set the IPA_SERVER_IP environment variable explicitly in the relevant container.
    
    Remove And Modify  securityContext Section

Unfortunately, none of these changes resolved the issue.

I'm wondering if there is any document or recommended approach for handling this scenario — especially how  run the pod successfully without privileged mode.

Below is my current manifest: 

```yaml

apiVersion: v1
kind: Service
metadata:
  name: freeipa
spec:
  selector:
    app: freeipa
  clusterIP: 10.233.5.115
  ports:
  - name: http
    port: 80
    targetPort: 80
  - name: https
    port: 443
    targetPort: 443
  - name: dns-tcp
    port: 53
    protocol: TCP
    targetPort: 53
  - name: dns-udp
    port: 53
    protocol: UDP
    targetPort: 53
  - name: kerberos-tcp
    port: 88
    protocol: TCP
    targetPort: 88
  - name: kerberos-udp
    port: 88
    protocol: UDP
    targetPort: 88
  - name: kpasswd-tcp
    port: 464
    protocol: TCP
    targetPort: 464
  - name: kpasswd-udp
    port: 464
    protocol: UDP
    targetPort: 464
  - name: ldap
    port: 389
    protocol: TCP
    targetPort: 389
  - name: ldaps
    port: 636
    protocol: TCP
    targetPort: 636
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: freeipa
spec:
  replicas: 1
  selector:
    matchLabels:
      app: freeipa
  template:
    metadata:
      labels:
        app: freeipa
    spec:
      containers:
        - name: freeipa-container
          image: freeipa/freeipa-server:rocky-9 # almalinux , centos  
          resources:
            requests:
              memory: "5Gi"
              cpu: "2"
            limits:
              memory: "5Gi"
              cpu: "2"
          imagePullPolicy: Always
          env:
            - name: IPA_SERVER_HOSTNAME
              value: freeipa.freetest.freeipame
            - name: IPA_SERVER_INSTALL_OPTS
              value: --unattended --realm=FREETEST.FREEIPAME --domain freetest.freeipame --ip-address=10.233.5.115 -p P@ssw0rd!@# -a P@ssw0rd!@# --no-forwarders --ssh-trust-dns --setup-dns --no-ntp 
          securityContext:
            capabilities:
              add: 
                - SYS_TIME
                - NET_BIND_SERVICE
            seccompProfile:
              type: Unconfined
          volumeMounts:
            - name: freeipa-server-data
              mountPath: /data
            - name: cgroups
              mountPath: /sys/fs/cgroup
            - name: run
              mountPath: /run
            - name: tmp
              mountPath: /tmp
            - mountPath: /dev/shm
              name: dshm
      volumes:
        - name: cgroups
          hostPath:
            path: /sys/fs/cgroup
        - name: run
          emptyDir:
            medium: Memory
        - name: tmp
          emptyDir:
            medium: Memory
        - name: dshm
          emptyDir:
            medium: Memory
            sizeLimit: 5Gi
        - name: freeipa-server-data
          persistentVolumeClaim:
            claimName: freeipa-server-data

```
Thanks in advance for your support and guidance

**Comment by adelton** [comment]
If you are after the solution as developed and maintained in https://github.com/iPenguin/bare-metal-kubernetes, you might get a faster help by filing the issue in that repo as the details of that host and Kubernetes setup might be important.

If you are after a FreeIPA Pod setup in your general Rocky Linux 9 Kubernetes 1.29 cluster, I'd suggest to start with https://github.com/freeipa/freeipa-container/blob/master/tests/freeipa-k3s.yaml, potentially with freeipa-k3s.yaml-*.patches from https://github.com/freeipa/freeipa-container/tree/master/tests applied, depending on your runtime. That describes a user-namespaced (`hostUsers: false`) setup we know works on Ubuntu 22.04 and 24.04, with CRI-O, user-namespaced docker, and as of today containerd 2.1-- that's what we test via GitHub Actions workflows multiple times a week. The choice of runtime might be important and the `UserNamespacesSupport` feature gate needs to get enabled and it was Alpha in Kubernetes 1.29, so not sure if that might spoil your chances. You might want to start with a testing Ubuntu 24.04 VM first, to get some working setup first.

I you hit problems, please file a new issue.

In general, `privileged: true` is not supported and will not work, as documented in the README (it applies to K8s-based setups as well), so consider that a dead end in your investigation. :-)
